GcodeHost Implementation

jamespearson04 opened the pull request:
Adds logic to ignore post command and full line comments whilst parsing Gcode from SD or FS through `GcodeHost::readNextCommand()`, as requested in #785.

If implemented, this renders the function `GcodeHost::isCommand()` redundant, as all it does is verify that `_currentCommand` isn't a full line comment, and trims any trailing comments from it.

This approach should reduce memory usage, as comments won't be written into _currentCommand to begin with. May also improve performance, as `GcodeHost::readNextCommand()` should only set `_currentCommand` to usable commands. (No extra calls required in the case of full line comments) #

Commit:
In readNextCommand() adds ';' to the "if" endline statement, if
character is ';', increments processedSize and CurrentPosition until
'/n' or '/r' is found. (effectively ignoring any comments received.)

If a comment is found, the command size is checked before marking
processing as finished, as full line comments would result in a command
size of 0, which would set HOST_STOP_STREAM as the next step.
With this check, it continues on to process the next line as though the
comment line wasn't there.

Commit:
In readNextCommand(), adds logic to skip through comments when parsing.
Works by treating ';' as an endline character. If it reaches this
character, it skips forward to the next endline character. If at this
point, the _currentCommand size is 0, it doesn't end processing, as this
suggests a full line comment. (Continues on to next line)

Commit:
Reorganized incrementing variables in SD to match FS and intial read
and increment sequence.

jamespearson04 commented:
Apologies for the messy commits, but should be understandable.

luc-github commented:
Thank you, I should be able to review the PR today

luc-github commented on esp3d/src/modules/gcode_host/gcode_host.cpp:
```diff
@@ -285,16 +292,23 @@ void GcodeHost::readNextCommand()
                         while(!((char)c == '\n') || ((char)c =='\r')){
                             _processedSize++;
                             _currentPosition++;
+                            c = FSfileHandle.read();
```
this is bad copy and paste should be  ` c = SDfileHandle.read();`
That make me realize  all part of  this code should be factorized between SD and FS as too much code duplication when the difference is actually only the  `c = SDfileHandle.read();` vs `c = FSfileHandle.read();` according define (SD_DEVICE /FILESYSTEM_FEATURE ) and _fsType (TYPE_SD_STREAM/TYPE_FS_STREAM)

jamespearson04 replied:
Absolutely, I see what you mean. I'll take care of that

luc-github commented:
the `GcodeHost::isCommand()` is now indeed a duplication but I think it can now be used to test if it is Host command (e.g: @pause) vs GCODE

jamespearson04 commented:
Got it, makes sense. Since I'm continuing to work on gcode hosting, this push isn't worth it as is, I'll convert it to a draft and update it.

Commit:
Added case and definitions for HOST_WAIT4_HEATING step. (Not yet
implemented)

Rewrote GCode/command line parsing in GcodeHost::readNextCommand() to
ignore empty lines and leading spaces - Previously empty lines would
set _step to HOST_STOP_STREAM. Now HOST_STOP_STREAM should only be set
in the case that the file read returns -1 (is this reliable at the end
of files?), or if a read is triggered when _processedSize => _totalSize.
(When the whole file has been read)

jamespearson04 committed:
Merge branch 'luc-github:3.0' into 3.0

Commit:
Merge remote-tracking branch 'refs/remotes/origin/3.0' into 3.0

Commit:
Further updated GcodeHost::readNextCommand(), was ending stream on full
line comments. Testing with serial monitor seems to indicate it's
working correctly now.

Commit:
Adds a missing increment to the _processedSize and _currentPosition
variables.

Commit:
-Implements line numbering and checksum for streamed gcode.
-Adds function for sending/authoring gcode commands
(GCodeHost::authorCommand()).
-Implements seperate timeout for heatup commands (timeout period TBD)
-add call to authorCommand in GCodeHost::pause(), GCodeHost::resume()
and GCodeHost::abort(), to send gcode commands M0, M108 and M112
respectively.
- Updated GcodeHost::resetCommandNumbering() to use authorCommand().

Commit:
-Commented out unnecessary HOST_STREAM_PAUSED case
-corrected starting line number

Commit:
-Added HOST_WAIT4_HEATING to switch.
-Fixed broken (else) if statement resulting in incorrect response message.

Commit:
Improved pause/resume handling

jamespearson04 commented:
Features added in this push so far:

- Checksum and line numbering
- Pause and resume w/ M0 and M108 commands
- Abort w/ M112 command

Still need to add line resend - If anyone knows where I can get some examples of error messages for missed lines etc please let me know

luc-github commented on esp3d/src/modules/gcode_host/gcode_host.cpp:
```diff
@@ -492,6 +579,8 @@ bool  GcodeHost::abort()
     log_esp3d("Aborting script");
     //TODO: what to do in addition ?
+    //Emergency stop for Marlin added (Maybe need a switch for other firmwares? -> should be mostly universal)
+    authorCommand("M112");
```
No M112 is not universal at least for GRBL and I am confused of the need of it actually this is a printer command - stop the stream does not necessary mean halt printer

jamespearson04 replied:
Will be removed. Would be useful to have the option to set custom gcode for the aborting - can be inserted here.

luc-github replied:
as mentioned in Discord IMHO this should be handled at ui level, do not try to handle such situation at stream level or it will be endless user case to manage, gcode host should only manage stream not printer

jamespearson04 replied:
Will want to define commands in ui of course, but will need to be handled as an addition to the stream. If we handle outside of gcodehost, might be making it difficult to update line number accordingly and handle ack, which could make pause unreliable.

I think whilst streaming, even whilst paused, gcodehost should handle all output to printer, except in the case of emergency stop. So if there's going to be any pause/resume/abort gcode, it needs to come through here.

 But I agree this needs changing. I think it wants to switch to streaming from a different source (pause/resume/abort scripts) before entering it's paused/resumed/aborted states. Of course if no scripts are set, this will be skipped and no extra commands will be sent to printer.

luc-github commented on esp3d/src/modules/gcode_host/gcode_host.cpp:
```diff
@@ -453,10 +534,16 @@ void GcodeHost::handle()
         break;
     case HOST_PAUSE_STREAM:
         //TODO pause stream
+        authorCommand("M0"); //M0 is unconditional stop in marlin
```
I am confused by this `authorCommand("M0"); `
 this command ask Marlin to pause and send message to host / lcd  to inform user
 so this command should be read by stream and interperted by Gcode host not sent by it to printer, this command is tipicaly used if GCODE file is read from SD, but the usage AFAK is not streamer send to printer to pause stream because streamer app is in charge of streaming, so if GCODE streamer see the command `M0` in the command it must send,  it actually do a pause, not send the `M0` 
 Or I miss a user case ?

jamespearson04 replied:
Would allow for the print to be resumed via the printer interface, but the pause routine would need to be updated to allow for it, which may hurt compatibility with other firmwares (would be waiting for ack like with M190 etc). At the moment it's more of a placeholder, since it shouldn't have any adverse effect.

I think the pause/resume/abort functions should have user configurable commands, since they can vary from printer to printer depending on things like parking the print head, and would allow for custom retraction scripts and the like. I'll replace them with defined strings for now with the hopes of having some text boxes added to the ui. That's a bit out of my skillset though, so would need some help with that part.

luc-github replied:
But if can resume from printer interface then how do  you ask gcode stream to resume?

jamespearson04 replied:
Would wait for ack like M190, so would just need a wait case in gcodehost::handle() that doesn't have a timeout. This would have to be enabled based on firmware, since this way of doing it would result in pauses being ignored on firmwares that don't support M0. Also requires marlin EMERGENCY_PARSER to be enabled if wish to resume via ESP, as M108 requires it.

luc-github commented on esp3d/src/modules/gcode_host/gcode_host.cpp:
```diff
@@ -453,10 +534,16 @@ void GcodeHost::handle()
     case HOST_RESUME_STREAM:
         //Any extra action to resume stream?
-        _step = HOST_READ_LINE;
+        authorCommand("M108"); //break and continue in marlin
```
same comment as `M0`

luc-github commented on esp3d/src/modules/gcode_host/gcode_host.cpp:
```diff
@@ -305,19 +371,91 @@ bool GcodeHost::isCommand()
     //TODO: what command do not need for ack ?
     return true;
 }
+void GcodeHost::authorCommand(String commandIn, String param)
```
why not use processCommand ?

jamespearson04 replied:
Allows the injection of commands like M112 regardless of if waiting for ack, and without juggling the contents of _currentCommand. Intended for use outside of normal gcode streaming, such as in pause/resume/abort. Could also be used for webui terminal -> serial, since this also wants to ignore ack and normal flow, though it likely works the same already.

luc-github replied:
but if using line numbering and check sum you cannot use not line numbering / no check sumbcommand in same time

jamespearson04 replied:
Need to verify, especially how this handles line numbers, but I believe you can use either Line no+checksum, or no line no+no checksum interchangeably.
Need to verify if commands without them still incremenet the expected line number so it doesn't break stream.

luc-github replied:
My mistake - I was misleading by my experience of M28/M29 which does not allows that, but it seems out of this usage you can switch numbered/checksum command and not numbered/checksum ccommands, I have verified on Marlin/Repetier/smoothieware and it is ok

luc-github replied:
but if command is no need for ack then in that case you just need to update https://github.com/luc-github/ESP3D/blob/3.0/esp3d/src/modules/gcode_host/gcode_host.cpp#L308-L312
```
bool GcodeHost::isAckNeeded()
{
    //TODO: what command do not need for ack ?
    return true;
}
```
to says if current command is `M112` return false

jamespearson04 replied:
True, M112 and M108 can be added there now. authorCommand was more of a placeholder for serial commands from terminal and other ui related functions, so that and associated calls will be replaced.
~~I'd missed how the ui uses gcodehost::push to pass commands in esp3doutput::dispatch, I'll update it to properly handle passed in commands.~~

jamespearson04 replied:
Also the current implementation of isAckNeeded doesn't help as it checks if an ack is needed after a command, not before

luc-github commented on esp3d/src/modules/gcode_host/gcode_host.cpp:
```diff
@@ -305,19 +371,91 @@ bool GcodeHost::isCommand()
+    }
+}
+
+void GcodeHost::authorCommand(String commandIn)
```
why not use `processCommand`, also if if different as `void GcodeHost::authorCommand(String commandIn, String param)` by just parameter why duplicate it ?

jamespearson04 replied:
Left in from some troubleshooting, my bad.

NoNamedCat replied:
Any news on this? I really need this feature. Tks a lot!

luc-github commented:
I did some comments - I did not finished to read all yet but I think there is confusion between:|
*  GCODE commands to be sent to host
* GCODE commands and Command to be interpreted by Host 
* GCODE commands to be sent to printer FW

jamespearson04 commented:
Currently @ commands haven't been implemented. I'm not too familiar with the ESP commands yet, but they appear to function properly, for instance ESP111 prints the IP address in the webui terminal, I'll try and get a screen recording of a test up.

jamespearson04 commented:
https://user-images.githubusercontent.com/26628667/180682955-90253256-b7ef-4c70-9294-99fba61b68ca.mp4

luc-github commented:
Did you read my comments in code?

Commit:
Removed extra space from log message.

Commit:
Removed the rest of the extra spaces.

marianomd commented:
Hi @jamespearson04, awesome work, what is the status of your branch? Is the feature working? What needs to be done @luc-github for this PR to get merged? Thanks a lot for the hard work.

luc-github commented:
need test all gcode provided by Marlin, check which gcode provide ack and which provide data, test each state: processing/busy/heating/error/ack/out of filament/ etc..
and handle accordingly

also handle a possible user commands like @ pause and @ continue


if ok, make all these parameters as variable according FW:  Marlin/Repetier/smoothieware

current pr is just at the early beginning, this is not simple task if handled properly

luc-github commented:
This feature is actually a project by itself

marianomd commented:
Thanks Luc for the detailed explanation. I really like the project, if I find time I will look into this.

luc-github commented:
sure, every contribution is welcome ^_^

jamespearson04 committed:
Merge branch 'luc-github:3.0' into 3.0
